hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationStateHolder.java

```diff
@@ -288,6 +320,25 @@ public int getId() {
         return MapDataSerializerHook.MAP_REPLICATION_STATE_HOLDER;
     }
 
+    @Override
+    public void setTarget(Address address) {
+        this.target = address;
+    }
+
+    private boolean mustWriteIndexInfos(Version clusterVersion) {
+        if (!getBuildInfo().isEnterprise() || clusterVersion.isGreaterOrEqual(V3_10)) {
+            return true;
+        }
+
+        ClusterService clusterService = operation.getNodeEngine().getClusterService();
+        Member targetMember = clusterService.getMember(target);
+        return targetMember.getVersion().asVersion().isEqualTo(V3_9) && clusterVersion.isEqualTo(V3_9);
```

Comment:
Maybe I'm just confused here but if the target member is e.g. EE 3.9.0 it will never read the index info but here we write it out anyway. Will this not affect the next object to read from the input stream on the 3.9.0 member?

Reply:
Maybe this will save your sanity:
- when writing data, the output's version is determined by the _local_ member, depending on cluster version, presence of `Versioned` interface, OS vs EE etc
- when reading data, the input's version comes across the network (or in general the source that serialized the object in the first place), so it is whatever the _remote_ member determined to be written out as the version.

In 3.9.x `MapReplicationStateHolder.readData` is like this:
```
        if (in.getVersion().isGreaterOrEqual(Versions.V3_9)) {
            // read it
        } else {
            mapIndexInfos = null;
        }
```

So:
- in 3.9, `MapReplicationOperation` is not `Versioned` -> a serialized object from 3.9 EE will report `UNKNOWN` as its version -> receiving-side 3.9 EE will not attempt to read any bytes
- in 3.10, `MapReplicationOperation` is `Versioned` -> its serialized output will carry the cluster version at the writing side across the wire to the receiving side -> receiving-side 3.9 EE will read the index info.

and no, it's not just you who is confused :)

Reply:
Oh, right, thanks for that, forgot the version is sent by the caller. Thanks!

Reply:
And in case the cluster version is 3.9, how come we don't send the index info to a 3.10 member? Is this because we the 3.10 members expect to receive indexes in the `PostJoinMapOperation`, same as from 3.9.x members?
Can this create a mismatch because the `mustReadMapIndexInfos` will see the sender version is 3.10 and return `true`?

Reply:
I think the decision tree in this methods deserves a few comments in the code. 
otherwise it's guaranteed in just a few weeks nobody - apart from Vassilis - will understand why the rules are like this.

Reply:
Yes, I have a hard time understanding it while I'm reviewing it :D

Reply:
I added a comment at the top of each class documenting the overall behavior and the decision tree; in general the idea is to get 3.10 members on cluster version 3.9 to behave as 3.9 members in a 3.9-only cluster.